### PR TITLE
Report code coverage that includes P tests

### LINKED ISSUE #2475: Measure Code Coverage for Pester
### Checklist

- [X] Feature request has a meaningful title
- [X] I have searched the existing issues. See [all issues](https://github.com/pester/Pester/issues?q=is%3Aissue)
- [X] I have tested using the latest version of Pester. See [Installation and update guide](https://pester.dev/docs/introduction/installation).

### Summary of the feature request

Pester is not enabling Code Coverage, probably because we are testing code coverage, and use P tests. Let's see if we can measure coverage somehow completely.

### How should it work?

_No response_

### PR DESCRIPTION
**Pull request by nohwnd**
Testing how we could measure CodeCoverage of Pester and P tests.  The code above runs, except for one test in mock, which is strange since we can Profile pester tests. Maybe it was some fluke. 

- [x] The coverage shows 28% which seems too low, most of our tests run in process, so we maybe are resetting something somewhere. Or not pointing the code at the right files.
- [x] This also needs to be made compatible with inlining the module.
- [x] And Frode pointed out that tracer might not be both the same type.
- [x] we also need finally to tear down
- [x] and print summary
- [x] We are probably not setting that one extra BP that we should be seetting, if all breakpoints are removed Set-Trace will turn off, check profiler what it does, so we can re-enable the test in Mock that removes just it's breakpoints.  https://github.com/nohwnd/Profiler/issues/67

- [x] ~Coverage tests fail but it is imho okay to skip~ these were mock tests, solved
- [x] coverage.ts tests break the coverage because they turn off tracing
- [x] this also higlights that when we profile Pester using PRofiler we are most likely not getting trace of all tests, so there are probably some other Perf wins hiding

image

Fix #2475

### CODE REVIEW

File: src/functions/Coverage.ps1 [hunk 1]
```diff
@@ -1122,9 +1122,18 @@ function Start-TraceScript ($Breakpoints) {
...
-        Set-PSDebug -Trace 1
+
+        # detect if code coverage is enabled throuh Pester tracer, and in that case just add us as a second tracer
+        if ([Pester.Tracing.Tracer]::ShouldRegisterTracer($tracer, <# overwrite: #> $false)) {
+            $patched = $false
+            $registered = $true
+            [Pester.Tracing.Tracer]::Register($tracer)
+        }
+        else {
+            $patched = $true
```
Comment (nohwnd) [inline src/functions/Coverage.ps1:1133]: > so I just needed to check in patch and unpatch not just for profiler tracer, but also for pester tracer.

@fflaten  this is what I meant by that. I should hide this under a Env variable probably, so we don't get into broken state after cancellation, as you pointed out.

Comment (nohwnd) [reply src/functions/Coverage.ps1:1133]: See few lines of the unchanged code above, it then makes more sense imho.

**Comment by nohwnd** [comment]
The error detection is sweet. Took me half evening yesterday to figure out where the code was stopping to properly trace, and this does it automatically: 

image

**Comment by nohwnd** [comment]
test.ps1 -cc

image

**Comment by nohwnd** [comment]
With inline. 

image

Not sure what is the difference, but it is small enough to ignore.

**Comment by nohwnd** [comment]
shows summary in UI now as well, it is not super helpful, but it is there. 

image

**Comment by johlju** [comment]
Fyi. The task `PublishCodeCoverageResults@2` does not (yet?) support line coverage, but if using v1 the `Pester.psm1` will be clickable and show coverage per line.

**Comment by nohwnd** [comment]
I tried that and I still don't see the uncovered lines in the breakdown. 
image

@fhnaseer is there a way to make the new code coverage view (two posts above) more detailed than just summary per file? we are producing jacoco output.

**Comment by johlju** [comment]
See example here where you can see it show hits and misses per line: https://dev.azure.com/dsccommunity/SqlServerDsc/_build/results?buildId=8834&view=codecoverage-tab

**Pull request merged**